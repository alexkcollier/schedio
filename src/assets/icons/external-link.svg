<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 23.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 64 64" style="enable-background:new 0 0 64 64;" xml:space="preserve">
<style type="text/css">
	.st0{display:none;fill:none;stroke:#004C97;stroke-width:4;stroke-miterlimit:10;}
	.st1{fill:none;stroke:#004C97;stroke-width:4;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
</style>
<path class="st0" d="M38,22H14c-2.21,0-4,1.79-4,4v24c0,2.21,1.79,4,4,4h24c2.21,0,4-1.79,4-4V26C42,23.79,40.21,22,38,22z"/>
<g>
	<line class="st1" x1="10" y1="54" x2="54" y2="10"/>
</g>
<line class="st1" x1="29.25" y1="10" x2="54" y2="10"/>
<line class="st1" x1="54" y1="34.75" x2="54" y2="10"/>
</svg>
