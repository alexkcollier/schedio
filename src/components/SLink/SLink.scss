.link {
  $highlight-spacing: 0.1em;

  color: color('spartan_blue');
  display: inline-block;
  position: relative;
  text-decoration: none;
  // Required for `z-index` to work properly on `::before`
  z-index: 1;

  // Add an obvious highlight
  &::before {
    background-color: color('green', 'lighter');
    border-radius: border-radius('small');
    // Move it up a bit. Uses bottom so highlight animates up
    // Percentages are favoured over ems as measurements are based on line-height.
    bottom: 10%;
    content: '';
    // Give it just a bit of visual padding
    left: -1 * $highlight-spacing;
    // Approximately align top with baseline
    height: 20%;
    position: absolute;
    transition: all duration('fast') easing();
    // Adjusts for left offset
    width: unquote('calc(100% + #{2 * $highlight-spacing})');
    z-index: -1;
  }

  // Animate the highlight
  &:hover {
    &::before {
      // Height is approximately from the bottom of `::before` to the top of the link text
      height: 80%;
    }
  }

  // add an icon to external links
  &--external {
    $icon-dimensions: 0.75em;
    $icon-margin: (1 - $icon-dimensions) / 2;
    // adds enough space before next character after link text
    margin-right: $icon-dimensions + $icon-margin;

    &::before {
      // increases highlight size to include icon and spacing
      width: unquote('calc(100% + #{2 * $highlight-spacing} + #{$icon-dimensions + $icon-margin})');
    }

    &::after {
      content: '';
      background: url('~@/assets/icons/external-link.svg') center center / contain no-repeat;
      height: $icon-dimensions;
      // put some horizontal space around the icon
      margin: 0 $icon-margin;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      width: $icon-dimensions;
    }
  }

  // Tone-down visited links
  &:visited {
    &::before {
      background-color: color('grey', 'light');
    }
  }
}
