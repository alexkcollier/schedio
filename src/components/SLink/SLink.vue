<script>
export default {
  name: 'SLink',

  props: {
    to: {
      type: [String, Object],
      required: true
    }
  },

  computed: {
    classList() {
      return { 'link--external': this.isExternalLink }
    },

    isExternalLink() {
      return typeof this.to === 'string' && /^(http(s)?|ftp):\/\//.test(this.to)
    },

    linkProperties() {
      if (this.isExternalLink) {
        return {
          is: 'a',
          href: this.to,
          target: '_blank',
          rel: 'noopener noreferrer'
        }
      }

      // handles internal links if `vue-router` is not present
      if (!this.$router && !this.isExternalLink) {
        return {
          is: 'a',
          href: this.to
        }
      }

      return {
        // nuxt compatibility
        is: this.$root.nuxt ? 'nuxt-link' : 'router-link',
        to: this.to
      }
    }
  },

  // using string template due to `eslint-plugin-vue`/vetur issues
  // can't disable `vue/component-is` rule in editor, but eslint cli picks it up properly
  template: `
    <component
      v-bind="[linkProperties, $attrs]"
      v-on="$listeners"
      :class="classList"
      class="link">
      <slot />
    </component>
  `
}
</script>
