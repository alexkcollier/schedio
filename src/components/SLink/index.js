import SLink from './SLink.vue';

export { SLink };

export default { install: Vue => Vue.component(SLink.name, SLink) };
