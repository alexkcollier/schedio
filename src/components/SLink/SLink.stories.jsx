import { SHeading } from '@/components/SHeading'
import { SLink } from '@/components/SLink'
import PropList from '@/docs/PropList'
import StoryContainer from '@/docs/StoryContainer'
import StoryLink from '@/docs/StoryLink'
import { text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/SLink', module)
  .addDecorator(withKnobs)
  .add('SLink', () => {
    const props = {
      to: text('Link to (required)', 'https://spartanbio.com', 'Props')
    }

    const linkText = text('Link text', 'Spartan Bio', 'Slots')

    // TODO: document visited/unvisited states.

    return {
      render() {
        return (
          <StoryContainer>
            <SHeading level="1">Links</SHeading>
            <p>
              Links are used for simple navigation. If you find yourself adding{' '}
              <code>href="#"</code> or <code>href=""javascript:void(0)"</code>, you should probably
              be using a <StoryLink to="Components/SButton" />.
            </p>
            <p>
              <code>SLink</code> will detect external and internal links automatically. It supports
              use with or without <code>vue-router</code>, and with <code>nuxt</code>.<code>a</code>
              , <code>router-link</code>, or <code>nuxt-link</code> will be used for the component
              as necessary.
            </p>

            <SHeading level="2">Example</SHeading>
            <p>
              Here's an external link to <SLink {...{ props }}>{linkText}</SLink>.
            </p>
            <p>
              Here's an{' '}
              <SLink to="/some-internal-page" on-click={e => e.preventDefault()}>
                internal link.
              </SLink>{' '}
              Don't worry; it's called with <code>@click.prevent</code>, so you won't leave the
              documentation..
            </p>

            <PropList component={SLink} />
          </StoryContainer>
        )
      }
    }
  })
