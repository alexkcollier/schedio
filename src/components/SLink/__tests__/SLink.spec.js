import { RouterLinkStub, shallowMount } from '@vue/test-utils';
import SLink from '@/components/SLink/SLink.vue';
import { axe, toHaveNoViolations } from 'jest-axe';
import icons from 'feather-icons/dist/icons.json';
const iconNames = Object.keys(icons);

expect.extend(toHaveNoViolations);

describe('SLink.vue', () => {
  let wrapper;
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { });

  beforeEach(() => {
    wrapper = shallowMount(SLink, {
      slots: {
        default: 'link text',
      },
      mocks: {
        $router: 'vue-router',
      },
      propsData: {
        to: 'https://spartanbio.com',
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('can display link text', async () => {
    expect(wrapper.vm.$slots.default).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('requires a url', () => {
    shallowMount(SLink);
    expect(errorSpy).toHaveBeenCalled();
  });

  it('detects external links and uses `a` tag for them', async () => {
    expect(wrapper.vm.isExternalLink).toBe(true);
    expect(wrapper.vm.componentIs).toBe('a');
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('respects `isExternal` prop', async () => {
    wrapper.setProps({ to: '/internal', isExternal: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isExternalLink).toBe(true);
    expect(wrapper.vm.componentIs).toBe('a');
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('can be set to `a`', async () => {
    const to = '/internal';

    wrapper.setProps({
      to,
      useAnchor: true,
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.vm.componentIs).toBe('a');
    expect(wrapper.attributes('href')).toBe(to);
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('uses `a` when no router is present', () => {
    const noRouter = shallowMount(SLink, {
      propsData: {
        to: '/about',
      },
    });

    expect(noRouter.vm.isExternalLink).toBe(false);
    expect(noRouter.vm.componentIs).toBe('a');
    expect(noRouter.html()).toMatchSnapshot();
  });

  it('can be plain', async () => {
    wrapper.setProps({ isPlain: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('can be light', async () => {
    wrapper.setProps({ isLight: true });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('can have an icon', async () => {
    wrapper.setProps({ icon: iconNames[0] });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('uses best practices for [target="_blank"]', () => {
    if (wrapper.element.target === '_blank') {
      expect(wrapper.element.rel).toBe('noopener noreferrer');
    }

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('detects internal links and uses a link component', async () => {
    wrapper.setProps({ to: '/about' });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isExternalLink).toBe(false);
    expect(wrapper.vm.componentIs).toBe('router-link');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('can use custom router components', () => {
    wrapper = shallowMount(SLink, {
      propsData: {
        to: '/about',
        linkComponent: 'custom-link',
      },
      mocks: {
        $router: 'vue-router',
      },
      stubs: {
        CustomLink: RouterLinkStub,
      },
    });

    expect(wrapper.vm.componentIs).toBe('custom-link');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
