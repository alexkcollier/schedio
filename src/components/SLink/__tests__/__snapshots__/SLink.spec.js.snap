// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLink.vue can be light 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link link--light">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue can be plain 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link link--plain">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue can display link text 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue can have an icon 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link">link text <sicon icon="activity" size="small" class="link__icon"></sicon>
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue detects external links and uses \`a\` tag for them 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue detects internal links and uses a link component 1`] = `
<a class="link">link text
  <!---->
  <!----></a>
`;

exports[`SLink.vue detects nuxt and uses \`nuxt-link\` instead of \`router-link\` 1`] = `
<a class="link">/about
  <!---->
  <!----></a>
`;

exports[`SLink.vue respects \`isExternal\` prop 1`] = `
<a href="/internal" target="_blank" rel="noopener noreferrer" class="link">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;

exports[`SLink.vue uses \`a\` when no router is present 1`] = `
<a href="/about" class="link">/about
  <!---->
  <!----></a>
`;

exports[`SLink.vue uses best practices for [target="_blank"] 1`] = `
<a href="https://spartanbio.com" target="_blank" rel="noopener noreferrer" class="link">link text
  <!---->
  <sicon icon="external-link" size="small" aria-label="External link" class="link__icon"></sicon>
</a>
`;
